src/referenced.cc

```diff
@@ -0,0 +1,166 @@
+/* Copyright 2017 Istio Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "referenced.h"
+
+#include "global_dictionary.h"
+#include "utils/md5.h"
+
+#include <set>
+#include <sstream>
+
+using ::istio::mixer::v1::ReferencedAttributes;
+
+namespace istio {
+namespace mixer_client {
+namespace {
+const char kDelimiter[] = "\0";
+const int kDelimiterLength = 1;
+const std::string kWordDelimiter = ":";
+}  // namespace
+
+bool Referenced::Fill(const ReferencedAttributes& reference) {
+  const std::vector<std::string>& global_words = GetGlobalWords();
+
+  for (const auto& match : reference.attribute_matches()) {
+    std::string name;
+    // name index from Mixer server needs to be decoded into string
+    // based on the global dictionary and per-message word list.
+    int idx = match.name();
+    if (idx >= 0) {
+      if ((unsigned int)idx >= global_words.size()) {
+        GOOGLE_LOG(ERROR) << "Global word index is too big: " << idx
+                          << " >= " << global_words.size();
+        return false;
+      }
+      name = global_words[idx];
+    } else {
+      // per-message index is negative, its format is:
+      //    per_message_idx = -(array_idx + 1)
+      idx = -idx - 1;
+      if (idx >= reference.words_size()) {
+        GOOGLE_LOG(ERROR) << "Per message word index is too big: " << idx
+                          << " >= " << reference.words_size();
+        return false;
+      }
+      name = reference.words(idx);
+    }
+
+    if (match.condition() == ReferencedAttributes::ABSENCE) {
+      absence_keys_.push_back(name);
+    } else if (match.condition() == ReferencedAttributes::EXACT) {
+      exact_keys_.push_back(name);
+    }
```

Comment:
Should we put in place some behavior such that the proxy will work correctly if mixer starts to return a REGEX match? In the current approach, I think if Mixer returns a REGEX match for an attribute, it will be completely ignored and thus lead to incorrect caching. It'd better to say "if a REGEX match is reported, we don't know what to do and the result isn't cached". It would be slow, but at least correct.

Reply:
Return false if received REGEX